chapters/programming_opencl_kernels.md

```diff
@@ -16,16 +16,25 @@ The OpenCL specification also enables optional *offline* compilation where the k
 
 To offset this portability problem, and to enable a richer language and compiler ecosystem, Khronos has defined a cross-vendor, portable intermediate program representation called [SPIR-V](https://www.khronos.org/spir/). An increasing number of OpenCL implementations are supporting ingestion of offline-compiled kernel programs in the SPIR-V format.
 
+SPIR-V enables independent innovation by the compiler and silicon communities. Compiler front ends that generate SPIR-V kernels that can be ingested and executed by any OpenCL driver that understands the SPIR-V format. For example the  [C++ for OpenCL](cpp_for_opencl.md) open source front-end and compilers for SYCL can generate SPIR-V code.
```

Comment:
I know you didn't write the preceding paragraph, but I think we should emphasize the "richer language and compiler ecosystem" more than the fix for the "portability problem".  IMHO the clear value of SPIR-V is that it enables kernels to be written in _any_ kernel programming language, not just in OpenCL, so long as the kernel programming language can be compiled to SPIR-V.

I'd add another sentence or two describing both the C++ for OpenCL work and the SYCL work - both of these are huge and important and deserve more than a passing mention.

Are there other languages we could mention also?  I know Julia can compile to SPIR-V, for example, though this is not currently used through the OpenCL backend.

Reply:
>Are there other languages we could mention also?

I think there is some Rust support for SPIR-V but I am not sure how official or active that it. Perhaps @AlexeySachkov can provide some more details?

> I know Julia can compile to SPIR-V, for example, though this is not currently used through the OpenCL backend.

Does Julia use compute version of SPIR-V? Then I think it is worth mentioning this because it doesn't matter what tools are used to generate it.

Ok, if we have some concrete suggestions what to add I am happy to add it in this PR, otherwise we could also create an issue to follow up on this later?

Reply:
> > Are there other languages we could mention also?
> I think there is some Rust support for SPIR-V but I am not sure how official or active that it. Perhaps @AlexeySachkov can provide some more details?

I know about the following usages of SPIR-V in other languages:
- Based on KhronosGroup/SPIRV-LLVM-Translator#504 by @maleadt it seems like there were some attempts to lower Julia down to SPIR-V
- Based on [geobacter-rs/SPIRV-LLVM-Translator](https://github.com/geobacter-rs/SPIRV-LLVM-Translator) fork of the translator by @DiamondLovesYou, it seems like there was an attempt to lower Rust into SPIR-V, but it seems like the project is not that active and now targets AMDGPU LLVM target, more info [here](https://github.com/geobacter-rs/geobacter#vulkanspirv)
- [thewilsonator/llvm-target-spirv](https://github.com/thewilsonator/llvm-target-spirv) by @thewilsonator was the attempt to embed the translator as a backend into LLVM for D language: [llvm-dev](https://lists.llvm.org/pipermail/llvm-dev/2017-May/112540.html)
- An [attempt](https://github.com/cheery/spirthon) to translate python bytecode to SPIR-V
- And of course, SYCL/DPC++: [intel/llvm](https://github.com/intel/llvm)

Reply:
Some details about the Julia use case: our [GPUCompiler.jl](https://github.com/JuliaGPU/GPUCompiler.jl) package supports compiling Julia code to SPIR-V, and that's actively being used in the [oneAPI.jl](https://github.com/JuliaGPU/oneAPI.jl) package for working with ~Intel~ oneAPI GPUs.

Reply:
Slightly OT, but

> Based on geobacter-rs/SPIRV-LLVM-Translator fork of the translator by @DiamondLovesYou, it seems like there was an attempt to lower Rust into SPIR-V, but it seems like the project is not that active and now targets AMDGPU LLVM target, more info here

You are right that that specific repo isn't active, but that's because that old translator isn't used anymore; instead, I built on the proper LLVM SPIRV target machine ARM created.

But you are right, due to time, I focus my time on Geobacter on AMDGPU so I don't have to deal with SPIRV's idiosyncrasies (resource descriptors! ugh) which LLVM doesn't have conducive means of representing. But Geobacter's principals are not intrinsically specific to any target.

Reply:
I have added info about Julia only since there is no stable path for Rust to SPIR-V. But it should be easy to update this if things change in the near future.